src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,415 @@
+ *
+ * An implementation of the LMNNFunction class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_LMNN_FUNCTION_IMPL_HPP
+#define MLPACK_METHODS_LMNN_FUNCTION_IMPL_HPP
+
+#include "lmnn_function.hpp"
+#include "constraints.hpp"
+
+#include <mlpack/core/math/make_alias.hpp>
+#include <mlpack/core/optimizers/function.hpp>
+
+namespace mlpack {
+namespace lmnn {
+
+template<typename MetricType>
+LMNNFunction<MetricType>::LMNNFunction(const arma::mat& dataset,
+                           const arma::Row<size_t>& labels,
+                           size_t k,
+                           double regularization,
+                           MetricType metric) :
+    dataset(math::MakeAlias(const_cast<arma::mat&>(dataset), false)),
+    labels(math::MakeAlias(const_cast<arma::Row<size_t>&>(labels), false)),
+    k(k),
+    metric(metric),
+    regularization(regularization)
+{
+  // Initialize the initial learning point.
+  initialPoint.eye(dataset.n_rows, dataset.n_rows);
+  // Initialize transformed dataset to base dataset.
+  transformedDataset = dataset;
+
+  // Initialize target neighbors & impostors.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+  constraint.Impostors(impostors);
+}
+
+//! Shuffle the dataset.
+template<typename MetricType>
+void LMNNFunction<MetricType>::Shuffle()
+{
+  arma::mat newDataset;
+  arma::Row<size_t> newLabels;
+
+  math::ShuffleData(dataset, labels, newDataset, newLabels);
+
+  math::ClearAlias(dataset);
+  math::ClearAlias(labels);
+
+  dataset = std::move(newDataset);
+  labels = std::move(newLabels);
+
+  // Re-calculate target neighbors as indices changed.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+}
+
+inline void Projection(arma::mat& iterate)
+{
+  // Projection.
+  arma::vec eigval;
+  arma::mat eigvec;
+
+  arma::eig_sym(eigval, eigvec, iterate);
+
+  arma::uvec ind = arma::find(eigval > 0);
+  arma::mat diagEigVal = arma::diagmat(eigval);
+  iterate = eigvec.cols(ind) * diagEigVal.submat(ind, ind) *
+      arma::trans(eigvec.cols(ind));
+}
+
+//! Evaluate cost over whole dataset.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates)
+{
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_elem)
+    Projection(iterate);
+
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  // Re-calculate impostors on transformed dataset.
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
```

Comment:
So, here we always recompute impostors, but we may not need to.  Let's say that `k = 3`, so for some point `p_i` and coordinates `M_t` we have `d_1(p_i)`, `d_2(p_i)`, and `d_3(p_i)`, which are the distances to the nearest impostor, second nearest impostor, and third nearest impostor.  Suppose that we also computed `d_4(p_i)`, the distance to the fourth nearest impostor.  When we take a step and get `M_{t + 1}`, we should be able to produce a lower bound on `d_4(p_i)` and an upper bound on `d_1(p_i)`, `d_2(p_i)`, and `d_3(p_i)`.  If that lower bound is still greater than any of the upper bounds, then the impostors cannot change and there is no need to recompute the impostors with a nearest neighbor search.

However I haven't yet had a chance to compute what those bounds actually would be.  I am thinking that I can probably express some bound with respect to the matrix norm || M_t - M_{t + 1} ||.  If you have a chance, do you want to give the bound a shot?

Reply:
If I understand this correctly, then we want to measure the change in impostors distance cause by each iteration and if it is still lower than lower bound (which we get by the distance of 
'k+1'th impostors ) then we don't need to recalculate impostors as they will still be the same. Please correct me if I am not correct here. 
If we can gain speedups, then I sure would like to work on this.

Reply:
That's correct.  The only thing is this is one of the more difficult speedups, so it's probably better to start with the simpler ideas and see how much they help.

Reply:
I spent some time with this and got a rough idea, I will be writing it down below. Please, Let me know if it sounds reasonable.

At every iteration the change in position of any data point is a measure of ( M_{t + 1} - M_{t} ) or more specifically  || ( M_{t + 1} - M_{t} ) *  p_{i} || . 
So what we can do is, keep an oldCoordinates and an oldImpostors (upto k + 1 impostors) matrices and then iterate over each impostors column (impostors corresponding to each data point, so total N columns) --- here within this iteration we will compare || ( M_{t + 1} - M_{t} ) *  p_{i} || ^ 2 for each impostor from i = 1 to i = k with  || M_{t} * ( p_{k + 1} - p_{i}) || ^ 2 and if the first term exceeds 2nd term for any of the i 's, then we will store the index of this data point in a Index vector ---- After performing the same operation for all the data points, we will have final Index vector (it will have indexes of those data points whose impostors are changed) and then we can calculate the impostors for just these points (Index vector may be a null vector too).

Reply:
Right, I think that that will work (I am taking L to be the same as M in your writing).  I did some derivation on the rest of the open whiteboard, attached below.  I want to write it up more formally but maybe it will give you an idea.  Basically I wanted to prove, if we know || L_{t + 1} - L_t ||, can we use that to bound which points could change impostors?  I have most of the result in the image but I don't know how legible it is.  I'll return to the idea tomorrow but I think I will end up with something roughly similar to the idea you have here.

![img_20180604_164437](https://user-images.githubusercontent.com/1845039/40940847-12c13e5c-6817-11e8-8747-89e7d53595a1.jpg)

If || L_{t + 1} - L_t || is sufficiently small, then we could know that impostors cannot change for any point at all, and thus we could avoid recalculation.

Alternately: when using SGD, there is "noise" in the gradient anyway by virtue of only calculating the gradient with respect to a few points at a time.  So if we were to say "we only recalculate impostors once every N passes through the data", or, "we only recalculate impostors when || L_{t + 1} - L_t || > a for some value a", that might still converge just fine.

I'll try and write the notes up into something more coherent tomorrow.  Sorry if they are not fully legible.